File: 02-LINKED-LISTS/PHP/implement-stack-using-queues.php
<?php


# https://leetcode.com/problems/implement-stack-using-queues/
class MyStack {
    private $queue;

    public function __construct() {
        $this->queue = new SplQueue();
    }

    public function push($x) {
        $this->queue->enqueue($x);
        $count = $this->queue->count();

        for ($i = 0; $i < $count - 1; $i++) {
            $this->queue->enqueue($this->queue->dequeue());
        }
    }

    public function pop() {
        return $this->queue->dequeue();
    }

    public function top() {
        return $this->queue->bottom();
    }

    public function empty() {
        return $this->queue->isEmpty();
    }
}

File: 02-LINKED-LISTS/Go/implement-stack-using-queues.go
package main

// https://leetcode.com/problems/implement-stack-using-queues/

type MyStack struct {
	stack [100]int
	size  int
}

func Constructor_() MyStack {
	return MyStack{
		stack: [100]int{},
		size:  0,
	}
}

func (this *MyStack) Push(x int) {
	this.stack[this.size] = x
	this.size++

}

func (this *MyStack) Pop() int {
	if this.size == 0 {
		return -1
	}

	this.size--
	return this.stack[this.size]
}

func (this *MyStack) Top() int {
	return this.stack[this.size-1]
}

func (this *MyStack) Empty() bool {
	return this.size == 0
}
